### PR TITLE
Port of latest grid weight generation script features

### PR DESCRIPTION
**Pull request by cjauvin**
This is a port of @julemai's latest grid weight generation script features into RavenPy. There are many things open to discussion!

### CODE REVIEW

**Comment by cjauvin** [comment]
New features:
* Support for 1d coords
* Support for NC input file

File: ravenpy/cli.py [hunk 1]
```diff
@@ -34,13 +34,23 @@ def cli():
     type=click.Tuple([str, str]),
     default=RoutingProductGridWeightImporter.VAR_NAMES,
     show_default=True,
-    help="Ordered variable names of longitude (x) and latitude (y) in the NetCDF file.",
+    help="(A) Variable name of 2D longitude and latitude variables in NetCDF (in this order). Example: 'lon,lat'. Or (B) Attribute name in input_file shapefile (option -i) that defines the index of the shape in NetCDF model output file (numbering needs to be [0 ... N-1]).",
```
Comment (julemai) [inline ravenpy/cli.py:37]: Revert to previous version as we now have the `routing-id-field` as separate option. Please leave "longitude (x)"  and "latitude (y)". It's easier for some people to refer to x/y.

**Comment by cjauvin** [comment]
Still to do (in the context of this PR):
* Implement the error reporting at the end of the generation script
* Add tests for `-s` and `-g` options

File: ravenpy/cli.py [hunk 2]
```diff
@@ -34,13 +34,23 @@ def cli():
...
     show_default=True,
+    help="Name of column in routing information shapefile containing unique key for each dataset.",
+)
+@click.option(
+    "-f",  # legacy script short option
+    "--netcdf-input-field",
+    type=str,
+    default=RoutingProductGridWeightImporter.NETCDF_INPUT_FIELD,
+    show_default=True,
+    help="Attribute name in input_file shapefile that defines the index of the shape in NetCDF model output file (numbering needs to be [0 ... N-1]).",
```
Comment (julemai) [inline ravenpy/cli.py:53]: Maybe: `input_file` --> `INPUT_FILE_PATH `

File: ravenpy/models/importers.py [hunk 3]
```diff
@@ -219,78 +219,82 @@ class RoutingProductGridWeightImporter:
...
+        elif input_file_path.suffix == ".shp":
+            self._input_data = geopandas.read_file(input_file_path)
+            self._input_is_netcdf = False
+        else:
+            raise ValueError(
+                "The input file must be a shapefile (.shp or .zip) or NetCDF"
+            )
+
+        routing_file_path = Path(routing_file_path)
+        if routing_file_path.suffix == ".zip":
```
Comment (julemai) [inline ravenpy/models/importers.py:271]: Should there be an `elif` for ".shp" and a `ValueError` if it's neither?

File: ravenpy/models/importers.py [hunk 4]
```diff
@@ -420,18 +379,191 @@ def extract(self) -> GridWeightsCommand:
...
+            if len(lon_dims) == 2 and len(lat_dims) == 2:
+                if lon_dims == self._dim_names:
+                    lon_var = np.transpose(lon_var)
+
+                if lat_dims == self._dim_names:
+                    lat_var = np.transpose(lat_var)
+
+            elif len(lon_dims) == 1 and len(lat_dims) == 1:
+                # Coord vars are 1d, make them 2d
+                lon_var_size = lon_var.size
```
Comment (julemai) [inline ravenpy/models/importers.py:445]: needed?

File: ravenpy/models/importers.py [hunk 5]
```diff
@@ -420,18 +379,191 @@ def extract(self) -> GridWeightsCommand:
...
+            for ishape in range(self._nlat):
+
+                idx = np.where(self._input_data[self._netcdf_input_field] == ishape)[0]
+                if len(idx) == 0:
+                    # print(
+                    #     "Polygon ID = {} not found in '{}'. Numbering of shapefile attribute '{}' needs to be [0 ... {}-1].".format(
+                    #         ishape, input_file, key_colname_model, nshapes
+                    #     )
+                    # )
+                    raise ValueError("Polygon ID not found.")
```
Comment (julemai) [inline ravenpy/models/importers.py:549]: Should we be a bit more talkative here telling the user which ID can't be found where and in which attribute column?

File: ravenpy/models/importers.py [hunk 6]
```diff
@@ -420,18 +379,191 @@ def extract(self) -> GridWeightsCommand:
...
+                if len(idx) > 1:
+                    # print(
+                    #     "Polygon ID = {} found multiple times in '{}' but needs to be unique. Numbering of shapefile attribute '{}' needs to be [0 ... {}-1].".format(
+                    #         ishape, input_file, key_colname_model, nshapes
+                    #     )
+                    # )
+                    raise ValueError("Polygon ID not unique.")
```
Comment (julemai) [inline ravenpy/models/importers.py:556]: More talkative?

File: tests/test_cli.py [hunk 7]
```diff
@@ -5,20 +5,55 @@
...
 
-    result = runner.invoke(generate_grid_weights, ds)
+    result = runner.invoke(generate_grid_weights, params)
 
     assert result.exit_code == 0
     assert not result.exception
     assert ":NumberHRUs 51" in result.output
     assert ":NumberGridCells 100" in result.output
     assert len(result.output.split("\n")) == 216
+
```
Comment (julemai) [inline tests/test_cli.py:23]: I think we need to test at least one weight that is derived..

File: tests/test_cli.py [hunk 8]
```diff
@@ -5,20 +5,55 @@
...
+    ]
+    params = map(str, params)
+
+    result = runner.invoke(generate_grid_weights, params)
+
+    assert result.exit_code == 0
+    assert not result.exception
+    assert ":NumberHRUs 51" in result.output
+    assert ":NumberGridCells 9801" in result.output
+    assert len(result.output.split("\n")) == 128
```
Comment (julemai) [inline tests/test_cli.py:42]: I think we need to test at least one weight that is derived..

File: tests/test_cli.py [hunk 9]
```diff
@@ -5,20 +5,55 @@
...
+        get_test_data("raven-routing-sample", "finalcat_hru_info.zip")[0],
+    ]
+    params = map(str, params)
+
+    result = runner.invoke(generate_grid_weights, params)
+
+    assert result.exit_code == 0
+    assert not result.exception
+    assert ":NumberHRUs 51" in result.output
+    assert ":NumberGridCells 810" in result.output
```
Comment (julemai) [inline tests/test_cli.py:58]: I think we need to test at least one weight that is derived..

File: ravenpy/models/importers.py (hunk 3 above)
Comment (cjauvin) [reply ravenpy/models/importers.py:271]: We could but since the CLI specifies that it must be a shapefile, if it's not we can just let the call to `geopandas.read_file` crash (this is a bit lazy of course).

Comment (Zeitsperre) [reply ravenpy/models/importers.py:271]: There isn't a special mimetype-like prepending needed for a shape. `geopandas.read_file()` should be able to handle them. Raising a `ValueError` will cause errors that otherwise wouldn't be raised (e.g. user sends a valid GeoJSON, GeoPackage, SQLite, etc.)

File: tests/test_cli.py [hunk 10]
```diff
@@ -5,20 +5,55 @@
...
+    result = runner.invoke(generate_grid_weights, params)
+
+    assert result.exit_code == 0
+    assert not result.exception
+    assert ":NumberHRUs 51" in result.output
+    assert ":NumberGridCells 9801" in result.output
+    assert len(result.output.split("\n")) == 128
+
+
+def test_generate_grid_weights_with_shp_input():
```
Comment (julemai) [inline tests/test_cli.py:45]: This test runs with the default `AREA_ERROR_THRESHOLD=0.05` which will result in several subbasins where weights do not sum up to 1.

I think it would be good to have a test with having `AREA_ERROR_THRESHOLD=0.5` and checking that one of those basins is adjusted correctly.

File: ravenpy/models/importers.py (hunk 4 above)
Comment (cjauvin) [reply ravenpy/models/importers.py:445]: Yes because I reassign `lon_var` to itself on the next line, which would cause the line after the next (`lat_var = ..`) to be buggy, without this memorized value.

File: ravenpy/models/importers.py (hunk 5 above)
Comment (cjauvin) [reply ravenpy/models/importers.py:549]: Yes absolutely, this is part of the additional error reporting that I want to implement in this PR. But I wanted to open it early, for you to see.

File: tests/test_cli.py (hunk 7 above)
Comment (cjauvin) [reply tests/test_cli.py:23]: If you want I can help you setting up to run (and write) tests.

File: ravenpy/models/importers.py [hunk 11]
```diff
@@ -219,78 +219,82 @@ class RoutingProductGridWeightImporter:
...
         # WGS 84 / North Pole LAEA Canada
-        self._shapes = self._shapes.to_crs(
+        self._routing_data = self._routing_data.to_crs(
             epsg=RoutingProductGridWeightImporter.CRS_CAEA
         )
 
-        self._shapes = self._shapes.drop_duplicates(self._hru_id_field).sort_values(
-            self._hru_id_field
-        )
+        self._routing_data = self._routing_data.drop_duplicates(
```
Comment (julemai) [inline ravenpy/models/importers.py:286]: I am unsure about dropping duplicates here...

File: tests/test_cli.py [hunk 12]
```diff
@@ -5,20 +5,55 @@
...
-    result = runner.invoke(generate_grid_weights, ds)
+    result = runner.invoke(generate_grid_weights, params)
 
     assert result.exit_code == 0
     assert not result.exception
     assert ":NumberHRUs 51" in result.output
     assert ":NumberGridCells 100" in result.output
     assert len(result.output.split("\n")) == 216
+
+
```
Comment (julemai) [inline tests/test_cli.py:24]: I think we need an additional test where we invoke multiple gauges to check that removing duplicates etc is working properly.

File: ravenpy/models/importers.py (hunk 11 above)
Comment (julemai) [reply ravenpy/models/importers.py:286]: I am suggesting an additional test below...

File: ravenpy/models/importers.py (hunk 3 above)
Comment (julemai) [reply ravenpy/models/importers.py:271]: ok. Was just wondering because we also distinguish between zip and shp for the `input_file_path`

File: ravenpy/models/importers.py (hunk 4 above)
Comment (julemai) [reply ravenpy/models/importers.py:445]: Absolutely!

File: ravenpy/models/importers.py (hunk 3 above)
Comment (cjauvin) [reply ravenpy/models/importers.py:271]: @Zeitsperre I'm pretty certaint that `geopandas.read_file` crashes if I do not prepend a `zip://` schema when required though.. you say it does not?

File: ravenpy/models/importers.py (hunk 11 above)
Comment (cjauvin) [reply ravenpy/models/importers.py:286]: You seem to be doing it in your own code though: https://github.com/julemai/GridWeightsGenerator/blob/main/derive_grid_weights.py#L453-L463

This is just the idiomatic numpy way of doing it.

File: ravenpy/models/importers.py (hunk 3 above)
Comment (Zeitsperre) [reply ravenpy/models/importers.py:271]: No, you are correct, it will crash. I'm referring to why it isn't necessary to perform similar formatting for other files.

File: ravenpy/models/importers.py (hunk 11 above)
Comment (julemai) [reply ravenpy/models/importers.py:286]: yes, but _after_ I have build the tree...

Comment (cjauvin) [reply ravenpy/models/importers.py:286]: ok so you think we should tolerate ROUTING_SHAPEFILE rows having the same HRU_ID?

Comment (julemai) [reply ravenpy/models/importers.py:286]: Yes, it's a long story. There is nothing to resolve this for now. I hate this too but...

File: tests/test_cli.py (hunk 10 above)
Comment (julemai) [reply tests/test_cli.py:95]: See: "test_generate_grid_weights_with_weight_rescaling()"

File: tests/test_cli.py (hunk 8 above)
Comment (julemai) [reply tests/test_cli.py:83]: Added
